### PR TITLE
Conditionally implement `miette::Diagnostic` for `pest::error:Error`

### LINKED ISSUE #582: Support miette::IntoDiagnostic for pest::error::Error<R>
It would be nice to be able to more easily use [miette](https://docs.rs/miette/) to format diagnostics generated by pest errors. Currently, since `pest::error::Error::message` is private, this cannot easily be done by third-party libraries that depend on both pest and miette without re-implementing much of the message formatting logic in pest. It would be helpful if either `message` was made public, or if there was a feature-gated `impl<R> IntoDiagnostic for pest::error::Error<R>` to make it easier to include pest errors in miette diagnostics.

Thank you! ♥

### PR DESCRIPTION
**Pull request by bobbbay**
Hides an implementation of [miette](lib.rs/miette)'s `Diagnostic` type under the `miette` feature flag. Also adds example usage to the `parens` example.

The only issue I have with this is that the Display of `error::Error` changes based on this flag. Since miette displays the error *and then* the diagnostic, the error would be printed twice: once in pest's form, and once in miette's form. Hnce, `error.rs:32-33` changes the display value of `Error` if the miette feature is enabled. In my opinion, it's not a clean solution, so I'd love to hear feedback.

Note: public API change: `pest::error::Error::formt(&self) -> String` is publicized, so users on the miette feature flag can still read Pest's errors if needed.

Closes #582

### CODE REVIEW

**Comment by CAD97** [comment]
> The only issue I have with this is that the Display of error::Error changes based on this flag.

The standard workaround for this would be a `.into_miette(self) -> impl Diagnostic` adapter. (`.as_miette(&self) -> impl '_ + Diagnostic` would be nicer, but having `'static` is necessary for normal use of miette. You *can* format non-`'static` diagnostics, but it's [more difficult](https://github.com/CAD97/tracing-filter/blob/v0.1/src/diagnostics.rs).)

File: pest/examples/parens.rs
Comment (CAD97) [inline pest/examples/parens.rs:94]: This would be better off keeping the behavior of `loop { try { ... } }`. That would be the same as the other `main`, e.g.

```suggestion
fn main() {
    loop {
        let mut line = String::new();

        print!("> ");
        io::stdout().flush().into_diagnostic()?;

        io::stdin().read_line(&mut line).unwrap();
        line.pop();

        let parse = ParenParser::parse(Rule::expr, &line);
        #[cfg(feature = "miette")]
        let parse = parse.map_err(<miette::Report as From>::from);
        match parse {
            Ok(pairs) => println!("{:?}", expr(pairs)),
            Err(e) => println!("\n{}", e),
        }
    }
}
```

(untested)

**Comment by bobbbay** [comment]
> The standard workaround for this would be a .into_miette(self) -> impl Diagnostic adapter.

Sure, but what would the `impl Diagnostic` be? It can't be `Self`, because it will have the other, proper formatting as Display. Unless there's another type - `MietteError` that implements `Diagnostic` and `From<Error>` - but that overcomplicates the design.

File: pest/examples/parens.rs
Comment (bobbbay) [reply pest/examples/parens.rs:94]: Unfortunately, it seems `Report`'s display only displays a message:

```
> x

Failed to parse at (1, 1)
>
```

Without any colors, or fancy source code formatting. The real fix would probably be to use an `impl ReportHandler`, but in order to see that, I need to acquire a `&mut Formatter`...

Comment (CAD97) [reply pest/examples/parens.rs:94]: You can use `{:?}` here to get the return-error-from-main formatting.

To get `&mut Formatter`, you `impl Display::fmt`.

**Comment by CAD97** [comment]
Personally, I don't think a non-exposed `MietteAdapter(Error)` (outside only sees `impl Diagnostic`) is a significant complication. The only difference is implementing `Display`/`Diagnostic` on `MietteAdapter` instead of `Error` directly.

For comparison, this is also the pattern used by `Path::display`.

File: pest/src/error.rs
Comment (tomtau) [inline pest/src/error.rs:556]: errors could also contain `path` and `continued_line` -- should those also be included in a `Diagnostic` (e.g. as additional labels) or would including those make the Miette default formatted print outs look strange?

Comment (bobbbay) [reply pest/src/error.rs:556]: Yes, it would make the miette-fornatted output strange. Adding it to the Display is fine, though.

File: pest/examples/parens.rs
Comment (CAD97) [inline pest/examples/parens.rs:83]: Is this true? (Does `impl Debug for Error` use the miette formatting? If not yet, doing so (when alt is false) seems like a good idea.)

**Review by CAD97 — CHANGES_REQUESTED** [review]
I'd want to see a snapshot-style test that shows one or more instances of the miette-formatted error before merging this.

(This could be as simple as `assert_eq!(format!("{it:?}"), "expected")` or as complex as (waiting for) #661 and using insta or expect_test.)

That said, it's about time to pass the buck on pressing the green button, so...

**Comment by CAD97** [comment]
(apparently I forgot to push the button to unassign the team when it chooses a load-balanced member)

File: pest/examples/parens.rs
Comment (bobbbay) [reply pest/examples/parens.rs:83]: `impl Debug for Error` does not, but `impl Debug for Report` does. And in this scenario, `e: Report`. I honestly don't thing that `Debug` for `Error` should print miette-styled messages, it should still be opt-in.

Comment (CAD97) [reply pest/examples/parens.rs:83]: I missed the conversion of the error into a `miette::Report` above.

**Comment by bobbbay** [comment]
Great!

(blocked by #666)

**Comment by prsabahrami** [comment]
Have you considered reopening this? I'm interested in working on this.

**Pull request closed without merging**
